// magmastream/dist/structures/Manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagerEventTypes = exports.PlayerStateEventTypes = exports.SearchPlatform = exports.UseNodeOptions = exports.Manager = void 0;
const tslib_1 = require("tslib");
const Utils_1 = require("./Utils");
const collection_1 = require("@discordjs/collection");
const events_1 = require("events");
const managerCheck_1 = tslib_1.__importDefault(require("../utils/managerCheck"));
const blockedWords_1 = require("../config/blockedWords");
const fs_1 = tslib_1.__importDefault(require("fs"));
const path_1 = tslib_1.__importDefault(require("path"));
/**
 * The main hub for interacting with Lavalink and using Magmastream,
 */
class Manager extends events_1.EventEmitter {
    /**
     * Attaches an event listener to the manager.
     * @param event The event to listen for.
     * @param listener The function to call when the event is emitted.
     * @returns The manager instance for chaining.
     */
    on(event, listener) {
        return super.on(event, listener);
    }
    /** The map of players. */
    players = new collection_1.Collection();
    /** The map of nodes. */
    nodes = new collection_1.Collection();
    /** The options that were set. */
    options;
    initiated = false;
    /**
     * Loads player states from the JSON file.
     * @param nodeId The ID of the node to load player states from.
     * @returns A promise that resolves when the player states have been loaded.
     */
    async loadPlayerStates(nodeId) {
        // Changed to async and added Promise<void>
        this.emit("debug", "[MANAGER] Loading saved players.");
        const node = this.nodes.get(nodeId);
        if (!node)
            throw new Error(`Could not find node: ${nodeId}`);
        const info = (await node.rest.getAllPlayers());
        const playerStatesDir = path_1.default.join(process.cwd(), "magmastream", "dist", "sessionData", "players");
        if (!fs_1.default.existsSync(playerStatesDir)) {
            fs_1.default.mkdirSync(playerStatesDir, { recursive: true });
        }
        const playerFiles = fs_1.default.readdirSync(playerStatesDir);
        /**
         * Converts a track from the Lavalink format to the Magmastream format.
         * @param song The track in the Lavalink format.
         * @returns The track in the Magmastream format.
         */
        const createTrackData = (song) => ({
            encoded: song.track,
            info: {
                identifier: song.identifier,
                isSeekable: song.isSeekable,
                author: song.author,
                length: song.duration,
                isrc: song.isrc,
                isStream: song.isStream,
                title: song.title,
                uri: song.uri,
                artworkUrl: song.artworkUrl,
                sourceName: song.sourceName,
            },
            pluginInfo: song.pluginInfo,
        });
        for (const file of playerFiles) {
            const filePath = path_1.default.join(playerStatesDir, file);
            if (!fs_1.default.existsSync(filePath)) {
                continue;
            }
            const data = fs_1.default.readFileSync(filePath, "utf-8");
            const state = JSON.parse(data);
            if (state && typeof state === "object" && state.guild && state.node.options.identifier === nodeId) {
                const lavaPlayer = info.find((player) => player.guildId === state.guild);
                if (!lavaPlayer) {
                    this.destroy(state.guild);
                    continue;
                }
                const playerOptions = {
                    guild: state.options.guild,
                    textChannel: state.options.textChannel,
                    voiceChannel: state.options.voiceChannel,
                    selfDeafen: state.options.selfDeafen,
                    volume: lavaPlayer.volume || state.options.volume,
                };
                this.emit("debug", `[MANAGER] Recreating player: ${state.guild} from saved file: ${JSON.stringify(state.options)}`);
                const player = this.create(playerOptions);
                if (!lavaPlayer.state.connected) {
                    player.connect();
                }
                const tracks = [];
                if (!lavaPlayer.track) {
                    if (state.queue.current !== null) {
                        for (const key in state.queue) {
                            if (!isNaN(Number(key)) && key !== "current" && key !== "previous" && key !== "manager") {
                                const song = state.queue[key];
                                tracks.push(Utils_1.TrackUtils.buildUnresolved(song, song.requester));
                            }
                        }
                        if (tracks.length > 0) {
                            player.queue.add(tracks);
                            if (!state.paused && lavaPlayer.state.connected)
                                player.play();
                        }
                        else {
                            const payload = {
                                reason: "finished",
                            };
                            node.queueEnd(player, state.queue.current, payload);
                        }
                    }
                    else {
                        if (state.queue.previous !== null) {
                            const payload = {
                                reason: "finished",
                            };
                            node.queueEnd(player, state.queue.previous, payload);
                        }
                        else {
                            this.destroy(state.guild);
                            continue;
                        }
                    }
                }
                else {
                    const currentTrack = state.queue.current;
                    tracks.push(Utils_1.TrackUtils.build(createTrackData(currentTrack), currentTrack.requester));
                    for (const key in state.queue) {
                        if (!isNaN(Number(key)) && key !== "current" && key !== "previous" && key !== "manager") {
                            const song = state.queue[key];
                            tracks.push(Utils_1.TrackUtils.buildUnresolved(song, song.requester));
                        }
                    }
                    player.queue.add(tracks);
                }
                if (state.paused)
                    player.pause(true);
                player.setTrackRepeat(state.trackRepeat);
                player.setQueueRepeat(state.queueRepeat);
                if (state.dynamicRepeat) {
                    player.setDynamicRepeat(state.dynamicRepeat, state.dynamicLoopInterval._idleTimeout);
                }
                if (state.isAutoplay && state?.data?.Internal_BotUser) {
                    player.setAutoplay(state.isAutoplay, state.data.Internal_BotUser);
                }
            }
        }
        this.emit("debug", "[MANAGER] Finished loading saved players.");
    }
    /**
     * Gets each player's JSON file
     * @param {string} guildId - The guild ID
     * @returns {string} The path to the player's JSON file
     */
    getPlayerFilePath(guildId) {
        // Get the directory path to where the player's JSON file will be saved
        const configDir = path_1.default.join(process.cwd(), "magmastream", "dist", "sessionData", "players");
        // Make sure the directory exists, create it if it doesn't
        if (!fs_1.default.existsSync(configDir)) {
            fs_1.default.mkdirSync(configDir, { recursive: true });
        }
        // Generate the full path to the player's JSON file
        const playerStateFilePath = path_1.default.join(configDir, `${guildId}.json`);
        return playerStateFilePath;
    }
    /**
     * Saves player states to the JSON file.
     * @param {string} guildId - The guild ID of the player to save
     */
    savePlayerState(guildId) {
        // Get the full path to the player's JSON file
        const playerStateFilePath = this.getPlayerFilePath(guildId);
        // Get the player instance from the manager's collection
        const player = this.players.get(guildId);
        // If the player does not exist or is disconnected, or the voice channel is not specified, do not save the player state
        if (!player || player.state === Utils_1.StateTypes.Disconnected || !player.voiceChannel) {
            // Clean up any inactive players
            return this.cleanupInactivePlayers();
        }
        // Serialize the player instance to avoid circular references
        const serializedPlayer = this.serializePlayer(player);
        // Write the serialized player state to the JSON file
        fs_1.default.writeFileSync(playerStateFilePath, JSON.stringify(serializedPlayer, null, 2), "utf-8");
        // Emit a debug event to indicate the player state has been saved
        this.emit("debug", `[MANAGER] Saving player: ${guildId} at location: ${playerStateFilePath}`);
    }
    /**
     * Serializes a Player instance to avoid circular references.
     * @param player The Player instance to serialize
     * @returns The serialized Player instance
     */
    serializePlayer(player) {
        const seen = new WeakSet();
        /**
         * Recursively serializes an object, avoiding circular references.
         * @param obj The object to serialize
         * @returns The serialized object
         */
        const serialize = (obj) => {
            if (obj && typeof obj === "object") {
                if (seen.has(obj))
                    return;
                seen.add(obj);
            }
            return obj;
        };
        const serializedPlayer = JSON.parse(JSON.stringify(player, (key, value) => {
            if (key === "filters" || key === "manager") {
                return null;
            }
            if (key === "queue") {
                return {
                    ...value,
                    current: value.current || null,
                };
            }
            return serialize(value);
        }));
        return serializedPlayer;
    }
    /**
     * Checks for players that are no longer active and deletes their saved state files.
     * This is done to prevent stale state files from accumulating on the file system.
     */
    cleanupInactivePlayers() {
        const playerStatesDir = path_1.default.join(process.cwd(), "magmastream", "dist", "sessionData", "players");
        // Create the directory if it does not exist
        if (!fs_1.default.existsSync(playerStatesDir)) {
            fs_1.default.mkdirSync(playerStatesDir, { recursive: true });
        }
        // Get the list of player state files
        const playerFiles = fs_1.default.readdirSync(playerStatesDir);
        // Get the set of active guild IDs from the manager's player collection
        const activeGuildIds = new Set(this.players.keys());
        // Iterate over the player state files
        for (const file of playerFiles) {
            // Get the guild ID from the file name
            const guildId = path_1.default.basename(file, ".json");
            // If the guild ID is not in the set of active guild IDs, delete the file
            if (!activeGuildIds.has(guildId)) {
                const filePath = path_1.default.join(playerStatesDir, file);
                fs_1.default.unlinkSync(filePath);
                this.emit("debug", `[MANAGER] Deleting inactive player: ${guildId}`);
            }
        }
    }
    /**
     * Returns the nodes that has the least load.
     * The load is calculated by dividing the lavalink load by the number of cores.
     * The result is multiplied by 100 to get a percentage.
     * @returns {Collection<string, Node>}
     */
    get leastLoadNode() {
        return this.nodes
            .filter((node) => node.connected)
            .sort((a, b) => {
            const aload = a.stats.cpu ? (a.stats.cpu.lavalinkLoad / a.stats.cpu.cores) * 100 : 0;
            const bload = b.stats.cpu ? (b.stats.cpu.lavalinkLoad / b.stats.cpu.cores) * 100 : 0;
            // Sort the nodes by their load in ascending order
            return aload - bload;
        });
    }
    /**
     * Returns the nodes that have the least amount of players.
     * Filters out disconnected nodes and sorts the remaining nodes
     * by the number of players in ascending order.
     * @returns {Collection<string, Node>} A collection of nodes sorted by player count.
     */
    get leastPlayersNode() {
        return this.nodes
            .filter((node) => node.connected) // Filter out nodes that are not connected
            .sort((a, b) => a.stats.players - b.stats.players); // Sort by the number of players
    }
    /**
     * Returns a node based on priority.
     * The nodes are sorted by priority in descending order, and then a random number
     * between 0 and 1 is generated. The node that has a cumulative weight greater than or equal to the
     * random number is returned.
     * If no node has a cumulative weight greater than or equal to the random number, the node with the
     * lowest load is returned.
     * @returns {Node} The node to use.
     */
    get priorityNode() {
        // Filter out nodes that are not connected or have a priority of 0
        const filteredNodes = this.nodes.filter((node) => node.connected && node.options.priority > 0);
        // Calculate the total weight
        const totalWeight = filteredNodes.reduce((total, node) => total + node.options.priority, 0);
        // Map the nodes to their weights
        const weightedNodes = filteredNodes.map((node) => ({
            node,
            weight: node.options.priority / totalWeight,
        }));
        // Generate a random number between 0 and 1
        const randomNumber = Math.random();
        // Initialize the cumulative weight to 0
        let cumulativeWeight = 0;
        // Loop through the weighted nodes and find the first node that has a cumulative weight greater than or equal to the random number
        for (const { node, weight } of weightedNodes) {
            cumulativeWeight += weight;
            if (randomNumber <= cumulativeWeight) {
                return node;
            }
        }
        // If no node has a cumulative weight greater than or equal to the random number, return the node with the lowest load
        return this.options.useNode === UseNodeOptions.LeastLoad ? this.leastLoadNode.first() : this.leastPlayersNode.first();
    }
    /**
     * Returns the node to use based on the configured `useNode` and `usePriority` options.
     * If `usePriority` is true, the node is chosen based on priority, otherwise it is chosen based on the `useNode` option.
     * If `useNode` is "leastLoad", the node with the lowest load is chosen, if it is "leastPlayers", the node with the fewest players is chosen.
     * If `usePriority` is false and `useNode` is not set, the node with the lowest load is chosen.
     * @returns {Node} The node to use.
     */
    get useableNodes() {
        return this.options.usePriority
            ? this.priorityNode
            : this.options.useNode === UseNodeOptions.LeastLoad
                ? this.leastLoadNode.first()
                : this.leastPlayersNode.first();
    }
    /**
     * Handles the shutdown of the process by saving all active players' states and optionally cleaning up inactive players.
     * This function is called when the process is about to exit.
     * It iterates through all players and calls {@link savePlayerState} to save their states.
     * Optionally, it also calls {@link cleanupInactivePlayers} to remove any stale player state files.
     * After saving and cleaning up, it exits the process.
     */
    async handleShutdown() {
        console.warn("\x1b[31m%s\x1b[0m", "MAGMASTREAM WARNING: Shutting down! Please wait, saving active players...");
        // Create an array of promises for saving player states
        const savePromises = Array.from(this.players.keys()).map((guildId) => {
            return new Promise((resolve) => {
                try {
                    this.savePlayerState(guildId);
                    resolve(); // Resolve immediately after calling savePlayerState
                }
                catch (error) {
                    console.error(`Error saving player state for guild ${guildId}:`, error);
                    throw error;
                }
            });
        });
        // Wait for all save operations to complete and check for errors
        const results = await Promise.allSettled(savePromises);
        const errors = results.filter((result) => result.status === "rejected");
        if (errors.length > 0) {
            console.error("`\x1b[31m%s\x1b[0m", `MAGMASTREAM ERROR: ${errors.length} player states failed to save.`);
        }
        // Clean up inactive players here
        this.cleanupInactivePlayers();
        console.warn("\x1b[32m%s\x1b[0m", "MAGMASTREAM INFO: Shutting down complete, exiting...");
        setTimeout(() => process.exit(errors.length > 0 ? 1 : 0), 100);
    }
    /**
     * Initiates the Manager class.
     * @param options
     * @param options.plugins - An array of plugins to load.
     * @param options.nodes - An array of node options to create nodes from.
     * @param options.autoPlay - Whether to automatically play the first track in the queue when the player is created.
     * @param options.autoPlaySearchPlatform - The search platform autoplay will use. Failback to Youtube if not found.
     * @param options.usePriority - Whether to use the priority when selecting a node to play on.
     * @param options.clientName - The name of the client to send to Lavalink.
     * @param options.defaultSearchPlatform - The default search platform to use when searching for tracks.
     * @param options.useNode - The strategy to use when selecting a node to play on.
     * @param options.trackPartial - The partial track search results to use when searching for tracks.
     * @param options.eventBatchDuration - The duration to wait before processing the collected player state events.
     * @param options.eventBatchInterval - The interval to wait before processing the collected player state events.
     */
    constructor(options) {
        super();
        process.on("SIGINT", async () => await this.handleShutdown());
        process.on("SIGTERM", async () => await this.handleShutdown());
        (0, managerCheck_1.default)(options);
        Utils_1.Structure.get("Player").init(this);
        Utils_1.Structure.get("Node").init(this);
        Utils_1.TrackUtils.init(this);
        if (options.trackPartial) {
            Utils_1.TrackUtils.setTrackPartial(options.trackPartial);
            delete options.trackPartial;
        }
        this.options = {
            plugins: [],
            nodes: [
                {
                    identifier: "default",
                    host: "localhost",
                    resumeStatus: false,
                    resumeTimeout: 1000,
                },
            ],
            autoPlay: true,
            usePriority: false,
            clientName: "Magmastream",
            defaultSearchPlatform: SearchPlatform.YouTube,
            autoPlaySearchPlatform: SearchPlatform.YouTube,
            useNode: UseNodeOptions.LeastPlayers,
            ...options,
        };
        if (this.options.plugins) {
            for (const [index, plugin] of this.options.plugins.entries()) {
                if (!(plugin instanceof Utils_1.Plugin))
                    throw new RangeError(`Plugin at index ${index} does not extend Plugin.`);
                plugin.load(this);
            }
        }
        if (this.options.nodes) {
            for (const nodeOptions of this.options.nodes)
                new (Utils_1.Structure.get("Node"))(nodeOptions);
        }
    }
    /**
     * Initiates the Manager.
     * @param clientId - The Discord client ID (required).
     * @returns The manager instance.
     */
    init(clientId) {
        if (this.initiated) {
            return this;
        }
        // Validate clientId
        if (typeof clientId !== "string" || !/^\d+$/.test(clientId)) {
            throw new Error('"clientId" must be a valid Discord client ID.');
        }
        // Set the validated clientId
        this.options.clientId = clientId;
        // Attempt to connect nodes
        for (const node of this.nodes.values()) {
            try {
                // Connect the node
                node.connect();
            }
            catch (err) {
                // Handle any errors that occur during the connection process
                this.emit("nodeError", node, err);
            }
        }
        // Set the initiated flag to true
        this.initiated = true;
        return this;
    }
    /**
     * Searches the enabled sources based off the URL or the `source` property.
     * @param query
     * @param requester
     * @returns The search result.
     */
    async search(query, requester) {
        const node = this.useableNodes;
        if (!node) {
            throw new Error("No available nodes.");
        }
        const _query = typeof query === "string" ? { query } : query;
        const _source = _query.source ?? this.options.defaultSearchPlatform;
        let search = _query.query;
        if (!/^https?:\/\//.test(search)) {
            search = `${_source}:${search}`;
        }
        this.emit("debug", `[MANAGER] Performing ${_source} search for: ${_query.query}`);
        try {
            const res = (await node.rest.get(`/v4/loadtracks?identifier=${encodeURIComponent(search)}`));
            if (!res) {
                throw new Error("Query not found.");
            }
            let searchData = [];
            let playlistData;
            switch (res.loadType) {
                case Utils_1.LoadTypes.Search:
                    searchData = res.data;
                    break;
                case Utils_1.LoadTypes.Track:
                    searchData = [res.data];
                    break;
                case Utils_1.LoadTypes.Playlist:
                    playlistData = res.data;
                    break;
            }
            const tracks = searchData.map((track) => Utils_1.TrackUtils.build(track, requester));
            let playlist = null;
            if (res.loadType === Utils_1.LoadTypes.Playlist) {
                playlist = {
                    name: playlistData.info.name,
                    playlistInfo: playlistData.pluginInfo,
                    requester: requester,
                    tracks: playlistData.tracks.map((track) => Utils_1.TrackUtils.build(track, requester)),
                    duration: playlistData.tracks.reduce((acc, cur) => acc + (cur.info.length || 0), 0),
                };
            }
            const result = {
                loadType: res.loadType,
                tracks,
                playlist,
            };
            if (this.options.replaceYouTubeCredentials) {
                const replaceCreditsURLs = ["youtube.com", "youtu.be"];
                const processTrack = (track) => {
                    if (!replaceCreditsURLs.some((url) => track.uri.includes(url)))
                        return track;
                    const { cleanTitle, cleanAuthor } = this.parseYouTubeTitle(track.title, track.author);
                    track.title = cleanTitle;
                    track.author = cleanAuthor;
                    return track;
                };
                if (result.loadType === Utils_1.LoadTypes.Playlist) {
                    result.playlist.tracks = result.playlist.tracks.map(processTrack);
                }
                else {
                    result.tracks = result.tracks.map(processTrack);
                }
            }
            this.emit("debug", `[MANAGER] Result ${_source} search for: ${_query.query}: ${JSON.stringify(result)}`);
            return result;
        }
        catch (err) {
            throw new Error(err);
        }
    }
    /**
     * Parses a YouTube title into a clean title and author.
     * @param title - The original title of the YouTube video.
     * @param originalAuthor - The original author of the YouTube video.
     * @returns An object with the clean title and author.
     */
    parseYouTubeTitle(title, originalAuthor) {
        // Remove "- Topic" from author and "Topic -" from title
        const cleanAuthor = originalAuthor.replace("- Topic", "").trim();
        title = title.replace("Topic -", "").trim();
        // Remove blocked words and phrases
        const escapedBlockedWords = blockedWords_1.blockedWords.map((word) => this.escapeRegExp(word));
        const blockedWordsPattern = new RegExp(`\\b(${escapedBlockedWords.join("|")})\\b`, "gi");
        title = title.replace(blockedWordsPattern, "").trim();
        // Remove empty brackets and balance remaining brackets
        title = title
            .replace(/[([{]\s*[)\]}]/g, "") // Empty brackets
            .replace(/^[^\w\d]*|[^\w\d]*$/g, "") // Leading/trailing non-word characters
            .replace(/\s{2,}/g, " ") // Multiple spaces
            .trim();
        // Remove '@' symbol before usernames
        title = title.replace(/@(\w+)/g, "$1");
        // Balance remaining brackets
        title = this.balanceBrackets(title);
        // Check if the title contains a hyphen, indicating potential "Artist - Title" format
        if (title.includes(" - ")) {
            const [artist, songTitle] = title.split(" - ").map((part) => part.trim());
            // If the artist part matches or is included in the clean author, use the clean author
            if (artist.toLowerCase() === cleanAuthor.toLowerCase() || cleanAuthor.toLowerCase().includes(artist.toLowerCase())) {
                return { cleanAuthor, cleanTitle: songTitle };
            }
            // If the artist is different, keep both parts
            return { cleanAuthor: artist, cleanTitle: songTitle };
        }
        // If no clear artist-title separation, return clean author and cleaned title
        return { cleanAuthor, cleanTitle: title };
    }
    /**
     * Balances brackets in a given string by ensuring all opened brackets are closed correctly.
     * @param str - The input string that may contain unbalanced brackets.
     * @returns A new string with balanced brackets.
     */
    balanceBrackets(str) {
        const stack = [];
        const openBrackets = "([{";
        const closeBrackets = ")]}";
        let result = "";
        // Iterate over each character in the string
        for (const char of str) {
            // If the character is an open bracket, push it onto the stack and add to result
            if (openBrackets.includes(char)) {
                stack.push(char);
                result += char;
            }
            // If the character is a close bracket, check if it balances with the last open bracket
            else if (closeBrackets.includes(char)) {
                if (stack.length > 0 && openBrackets.indexOf(stack[stack.length - 1]) === closeBrackets.indexOf(char)) {
                    stack.pop();
                    result += char;
                }
            }
            // If it's neither, just add the character to the result
            else {
                result += char;
            }
        }
        // Close any remaining open brackets by adding the corresponding close brackets
        while (stack.length > 0) {
            const lastOpen = stack.pop();
            result += closeBrackets[openBrackets.indexOf(lastOpen)];
        }
        return result;
    }
    /**
     * Escapes a string by replacing special regex characters with their escaped counterparts.
     * @param string - The string to escape.
     * @returns The escaped string.
     */
    escapeRegExp(string) {
        // Replace special regex characters with their escaped counterparts
        return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    }
    /**
     * Decodes an array of base64 encoded tracks and returns an array of TrackData.
     * Emits a debug event with the tracks being decoded.
     * @param tracks - An array of base64 encoded track strings.
     * @returns A promise that resolves to an array of TrackData objects.
     * @throws Will throw an error if no nodes are available or if the API request fails.
     */
    decodeTracks(tracks) {
        this.emit("debug", `[MANAGER] Decoding tracks: ${JSON.stringify(tracks)}`);
        return new Promise(async (resolve, reject) => {
            const node = this.nodes.first();
            if (!node)
                throw new Error("No available nodes.");
            const res = (await node.rest.post("/v4/decodetracks", JSON.stringify(tracks)).catch((err) => reject(err)));
            if (!res) {
                return reject(new Error("No data returned from query."));
            }
            return resolve(res);
        });
    }
    /**
     * Decodes a base64 encoded track and returns a TrackData.
     * @param track - The base64 encoded track string.
     * @returns A promise that resolves to a TrackData object.
     * @throws Will throw an error if no nodes are available or if the API request fails.
     */
    async decodeTrack(track) {
        const res = await this.decodeTracks([track]);
        // Since we're only decoding one track, we can just return the first element of the array
        return res[0];
    }
    /**
     * Creates a player or returns one if it already exists.
     * @param options The options to create the player with.
     * @returns The created player.
     */
    create(options) {
        if (this.players.has(options.guild)) {
            return this.players.get(options.guild);
        }
        // Create a new player with the given options
        this.emit("debug", `[MANAGER] Creating new player with options: ${JSON.stringify(options)}`);
        return new (Utils_1.Structure.get("Player"))(options);
    }
    /**
     * Returns a player or undefined if it does not exist.
     * @param guild The guild ID of the player to retrieve.
     * @returns The player if it exists, undefined otherwise.
     */
    get(guild) {
        return this.players.get(guild);
    }
    /**
     * Destroys a player if it exists and cleans up inactive players.
     * @param guild - The guild ID of the player to destroy.
     * @returns {void}
     * @emits {debug} - Emits a debug message indicating the player is being destroyed.
     */
    destroy(guild) {
        // Emit debug message for player destruction
        this.emit("debug", `[MANAGER] Destroying player: ${guild}`);
        // Remove the player from the manager's collection
        this.players.delete(guild);
        // Clean up any inactive players
        this.cleanupInactivePlayers();
    }
    /**
     * Creates a new node or returns an existing one if it already exists.
     * @param options - The options to create the node with.
     * @returns The created node.
     */
    createNode(options) {
        // Check if the node already exists in the manager's collection
        if (this.nodes.has(options.identifier || options.host)) {
            // Return the existing node if it does
            return this.nodes.get(options.identifier || options.host);
        }
        // Emit a debug event for node creation
        this.emit("debug", `[MANAGER] Creating new node with options: ${JSON.stringify(options)}`);
        // Create a new node with the given options
        return new (Utils_1.Structure.get("Node"))(options);
    }
    /**
     * Destroys a node if it exists. Emits a debug event if the node is found and destroyed.
     * @param identifier - The identifier of the node to destroy.
     * @returns {void}
     * @emits {debug} - Emits a debug message indicating the node is being destroyed.
     */
    destroyNode(identifier) {
        const node = this.nodes.get(identifier);
        if (!node)
            return;
        this.emit("debug", `[MANAGER] Destroying node: ${identifier}`);
        node.destroy();
        this.nodes.delete(identifier);
    }
    /**
     * Sends voice data to the Lavalink server.
     * @param data
     */
    async updateVoiceState(data) {
        if ("t" in data && !["VOICE_STATE_UPDATE", "VOICE_SERVER_UPDATE"].includes(data.t))
            return;
        const update = "d" in data ? data.d : data;
        if (!update || (!("token" in update) && !("session_id" in update)))
            return;
        const player = this.players.get(update.guild_id);
        if (!player)
            return;
        this.emit("debug", `[MANAGER] Updating voice state: ${JSON.stringify(update)}`);
        if ("token" in update) {
            player.voiceState.event = update;
            const { sessionId, event: { token, endpoint }, } = player.voiceState;
            await player.node.rest.updatePlayer({
                guildId: player.guild,
                data: { voice: { token, endpoint, sessionId } },
            });
            return;
        }
        if (update.user_id !== this.options.clientId)
            return;
        if (update.channel_id) {
            if (player.voiceChannel !== update.channel_id) {
                this.emit("playerMove", player, player.voiceChannel, update.channel_id);
            }
            player.voiceState.sessionId = update.session_id;
            player.voiceChannel = update.channel_id;
            return;
        }
        this.emit("playerDisconnect", player, player.voiceChannel);
        player.voiceChannel = null;
        player.voiceState = Object.assign({});
        player.destroy();
        return;
    }
}
exports.Manager = Manager;
var UseNodeOptions;
(function (UseNodeOptions) {
    UseNodeOptions["LeastLoad"] = "leastLoad";
    UseNodeOptions["LeastPlayers"] = "leastPlayers";
})(UseNodeOptions || (exports.UseNodeOptions = UseNodeOptions = {}));
var SearchPlatform;
(function (SearchPlatform) {
    SearchPlatform["YouTubeMusic"] = "ytmsearch";
    SearchPlatform["YouTube"] = "ytsearch";
    SearchPlatform["Spotify"] = "spsearch";
    SearchPlatform["Jiosaavn"] = "jssearch";
    SearchPlatform["SoundCloud"] = "scsearch";
    SearchPlatform["Deezer"] = "dzsearch";
    SearchPlatform["Tidal"] = "tdsearch";
    SearchPlatform["AppleMusic"] = "amsearch";
    SearchPlatform["Bandcamp"] = "bcsearch";
})(SearchPlatform || (exports.SearchPlatform = SearchPlatform = {}));
var PlayerStateEventTypes;
(function (PlayerStateEventTypes) {
    PlayerStateEventTypes["AutoPlayChange"] = "playerAutoplay";
    PlayerStateEventTypes["ConnectionChange"] = "playerConnection";
    PlayerStateEventTypes["RepeatChange"] = "playerRepeat";
    PlayerStateEventTypes["PauseChange"] = "playerPause";
    PlayerStateEventTypes["QueueChange"] = "queueChange";
    PlayerStateEventTypes["TrackChange"] = "trackChange";
    PlayerStateEventTypes["VolumeChange"] = "volumeChange";
    PlayerStateEventTypes["ChannelChange"] = "channelChange";
    PlayerStateEventTypes["PlayerCreate"] = "playerCreate";
    PlayerStateEventTypes["PlayerDestroy"] = "playerDestroy";
})(PlayerStateEventTypes || (exports.PlayerStateEventTypes = PlayerStateEventTypes = {}));
var ManagerEventTypes;
(function (ManagerEventTypes) {
    ManagerEventTypes["Debug"] = "debug";
    ManagerEventTypes["NodeCreate"] = "nodeCreate";
    ManagerEventTypes["NodeDestroy"] = "nodeDestroy";
    ManagerEventTypes["NodeConnect"] = "nodeConnect";
    ManagerEventTypes["NodeReconnect"] = "nodeReconnect";
    ManagerEventTypes["NodeDisconnect"] = "nodeDisconnect";
    ManagerEventTypes["NodeError"] = "nodeError";
    ManagerEventTypes["NodeRaw"] = "nodeRaw";
    ManagerEventTypes["PlayerCreate"] = "playerCreate";
    ManagerEventTypes["PlayerDestroy"] = "playerDestroy";
    ManagerEventTypes["PlayerStateUpdate"] = "playerStateUpdate";
    ManagerEventTypes["PlayerMove"] = "playerMove";
    ManagerEventTypes["PlayerDisconnect"] = "playerDisconnect";
    ManagerEventTypes["QueueEnd"] = "queueEnd";
    ManagerEventTypes["SocketClosed"] = "socketClosed";
    ManagerEventTypes["TrackStart"] = "trackStart";
    ManagerEventTypes["TrackEnd"] = "trackEnd";
    ManagerEventTypes["TrackEndReason"] = "trackEndReason";
    ManagerEventTypes["TrackEndReasonRaw"] = "trackEndReasonRaw";
    ManagerEventTypes["TrackEndReasonData"] = "trackEndReasonData";
    ManagerEventTypes["TrackStuck"] = "trackStuck";
    ManagerEventTypes["TrackError"] = "trackError";
    ManagerEventTypes["SegmentsLoaded"] = "segmentsLoaded";
    ManagerEventTypes["SegmentSkipped"] = "segmentSkipped";
    ManagerEventTypes["ChapterStarted"] = "chapterStarted";
    ManagerEventTypes["ChaptersLoaded"] = "chaptersLoaded";
})(ManagerEventTypes || (exports.ManagerEventTypes = ManagerEventTypes = {}));
